### PR TITLE
feat: add @netlify/vite-plugin-react-router to support React Router 7

### PR DESCRIPTION
**Pull request by serhalp**
## Description

Remix has converged with React Router and the "next version" of Remix was released as React Router 7.

This PR adds a new package to support React Router 7: `@netlify/vite-plugin-react-router`. The naming convention is different as this is purely a Vite plugin (previous "adapters" also supported the Remix Classic Compiler), so we were able to follow the [documented naming convention](https://vite.dev/guide/api-plugin#conventions).

## Future work

- This does not yet add support for edge rendering.
- An upgrade/migration guide from Remix 2 is coming soon
- The plugin source is 95% duplicated from `@netlify/remix-adapter`. We should probably refactor to extract the shared logic.

## Related Tickets & Documents

FRB-1511

Supersedes https://github.com/netlify/remix-compute/pull/471

See also https://github.com/netlify/remix-template/issues/157

## QA Instructions, Screenshots, Recordings

This PR adds extensive e2e tests.

I also added a new React Router Netlify template here: https://github.com/netlify/react-router-template/pull/1/. It uses this new plugin, and you can see the successful deploy preview there.

### CODE REVIEW

File: packages/remix-adapter/package.json [hunk 1]
```diff
@@ -67,14 +67,12 @@
     "@remix-run/react": "^2.12.0",
     "@types/react": "^18.0.27",
     "@types/react-dom": "^18.0.10",
-    "lambda-tester": "^4.0.1",
```
Comment (serhalp) [inline packages/remix-adapter/package.json:70]: this was unused

File: packages/remix-adapter/package.json [hunk 2]
```diff
@@ -67,14 +67,12 @@
...
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "tsup": "^8.0.2",
-    "vite": "^5.1.3"
+    "vite": "^5.4.11"
   },
   "peerDependencies": {
-    "@netlify/functions": "^2.8.1",
```
Comment (serhalp) [inline packages/remix-adapter/package.json:77]: I don't think this was ever relevant, or at least it isn't currently

File: packages/remix-adapter/package.json [hunk 3]
```diff
@@ -67,14 +67,12 @@
     "@remix-run/react": "^2.12.0",
     "@types/react": "^18.0.27",
     "@types/react-dom": "^18.0.10",
-    "lambda-tester": "^4.0.1",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "tsup": "^8.0.2",
-    "vite": "^5.1.3"
+    "vite": "^5.4.11"
```
Comment (serhalp) [inline packages/remix-adapter/package.json:73]: I had to upgrade this to resolve some type issues caused by having multiple versions of vite in the workspace.

File: packages/vite-plugin-react-router/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+The React Router Adapter for Netlify allows you to deploy your [React Router](https://reactrouter.com) app to
+[Netlify Functions](https://docs.netlify.com/functions/overview/).
+
+To deploy a React Router 7+ site to Netlify, install this package:
+
+## How to use
+
+```sh
+npm install @netlify/vite-plugin-react-router
```
Comment (serhalp) [inline packages/vite-plugin-react-router/README.md:11]:
```suggestion
npm install --save-dev @netlify/vite-plugin-react-router
```

File: packages/vite-plugin-react-router/src/plugin.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import type { Plugin, ResolvedConfig } from 'vite'
+import { mkdir, writeFile } from 'node:fs/promises'
+import { join, relative, sep } from 'node:path'
+import { sep as posixSep } from 'node:path/posix'
+import { version, name } from '../package.json'
+
+const NETLIFY_FUNCTIONS_DIR = '.netlify/functions-internal'
```
Comment (serhalp) [inline packages/vite-plugin-react-router/src/plugin.ts:7]: oh, we should probably take 15 mins to update this to use the Frameworks API instead of spreading this to a brand "new" adapter

File: packages/vite-plugin-react-router/src/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+          if (typeof config.build?.rollupOptions?.input === 'string') {
+            config.build.rollupOptions.input = {
+              [FUNCTION_HANDLER_CHUNK]: FUNCTION_HANDLER_MODULE_ID,
+              index: config.build.rollupOptions.input,
+            }
+            if (config.build.rollupOptions.output && !Array.isArray(config.build.rollupOptions.output)) {
+              config.build.rollupOptions.output.entryFileNames = '[name].js'
+            }
+          }
+        }
```
Comment (serhalp) [inline packages/vite-plugin-react-router/src/plugin.ts:74]: I think we can probably simplify this mess a lot. Avoid scope creep, but maybe timebox 10 mins on rewriting this part from scratch. Remix's own template just did this and it worked: https://github.com/remix-run/react-router-templates/blob/ff303a5bd4c742104c601dc5eda4f825d10739b8/netlify/vite.config.ts#L8-L14.

File: packages/vite-plugin-react-router/src/plugin.ts (hunk 5 above)
Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts:7]: ... I think it's literally just changing this one path

File: tests/e2e/react-router-user-journeys.spec.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
+import { expect, test } from './support/fixtures'
+
+test.describe('React Router user journeys', () => {
```
Comment (serhalp) [inline tests/e2e/react-router-user-journeys.spec.ts:3]: These are almost identical to the Remix e2e tests. We only changed a few strings like "Welcome to Remix", etc.

File: packages/vite-plugin-react-router/src/plugin.ts (hunk 6 above)
Comment (mrstork) [reply packages/vite-plugin-react-router/src/plugin.ts:74]: Let's do it!

File: tests/e2e/fixtures/react-router-serverless-site/netlify.toml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# TODO(serhalp) Remove entire file when we've added RR7 framework detection
```
Comment (mrstork) [inline tests/e2e/fixtures/react-router-serverless-site/netlify.toml:1]: Include a link please

File: tests/e2e/fixtures/react-router-serverless-site/vite.config.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+import tsconfigPaths from 'vite-tsconfig-paths'
+import netlifyPlugin from '@netlify/vite-plugin-react-router'
+
+export default defineConfig({
+  css: {
+    postcss: {
+      plugins: [tailwindcss, autoprefixer],
+    },
+  },
+  plugins: [reactRouter(), netlifyPlugin(), tsconfigPaths()],
```
[the remark this answers is not in the document]
Comment (serhalp) [reply tests/e2e/fixtures/react-router-serverless-site/vite.config.ts:14]: For posterity: this is what I was referencing in https://github.com/netlify/remix-compute/pull/472#discussion_r1879053633

File: packages/vite-plugin-react-router/src/plugin.ts (hunk 5 above)
Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts:7]: oh yeah, I always forget that _technically_ changing to use the Frameworks API is a breaking change because it won't be supported for manual deploys with [netlify-cli < 17.31.0](https://github.com/netlify/cli/pull/6737)

which is to say we should definitely use it in this brand new package

Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts:7]: Welp, I just changed the path and it just worked ✅

File: packages/vite-plugin-react-router/src/plugin.ts (hunk 6 above)
Comment (serhalp) [reply packages/vite-plugin-react-router/src/plugin.ts]: Woo, this sorta just worked too. Ultimately it ends up being roughly the same number of lines, but I think it's easier to understand and no longer requires a bunch of TODOs and NOTEs. :)

We should also be able to backport this refactor to the Remix plugins later.

**Review by mrstork — APPROVED** [review]
Nice!
